File: src/main/java/me/oriharel/machinery/upgrades/SpeedUpgrade.java
package me.oriharel.machinery.upgrades;

import me.oriharel.machinery.Machinery;
import me.oriharel.machinery.machine.MachineResourceGetProcess;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.YamlConfiguration;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import java.util.AbstractMap;
import java.util.Map;
import java.util.stream.Collectors;

public class SpeedUpgrade extends AbstractUpgrade {

    protected Map<Integer, Integer> costs;

    public SpeedUpgrade(int level) {
        super(level);
        runOnlyOnProcessStart = true;
        costs = null;
    }

    @Override
    public void applyUpgradeModifier(MachineResourceGetProcess mineProcess) {
        YamlConfiguration config = Machinery.getInstance().getFileManager().getConfig("upgrades.yml").get();
        ScriptEngineManager mgr = new ScriptEngineManager();
        ScriptEngine engine = mgr.getEngineByName("JavaScript");
        // period is in ticks. 20 ticks is a second
        int basePeriod = config.getInt("speed.basePeriod") * 20;
        if (level == 1) {
            mineProcess.setMinePeriod(basePeriod);
            return;
        }
        try {
            mineProcess.setMinePeriod((Integer) engine.eval(config.getString("speed.calculate").replaceAll("\\{basePeriod}", String.valueOf(basePeriod)).replaceAll(
                    "\\{level}",
                    String.valueOf(level))));
        } catch (ScriptException e) {
            Bukkit.getLogger().severe("Failed to calculate mine period!");
            Bukkit.getLogger().severe(e.getMessage());
        }
    }

    @Override
    public String getUpgradeName() {
        return "Speed Upgrade";
    }

    @Override
    public Map<Integer, Integer> getCosts() {
        if (costs == null) costs =
                configLoad.getConfigurationSection("speed.costs").getValues(false).entrySet().stream().map(entry -> new AbstractMap.SimpleEntry<>(Integer.valueOf(entry.getKey()), Integer.parseInt((String) entry.getValue()))).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
        return costs;
    }

    @Override
    public UpgradeType getUpgradeType() {
        return UpgradeType.SPEED;
    }

    @Override
    public String toString() {
        return "SpeedUpgrade{" +
                "level=" + level +
                ", runOnlyOnProcessStart=" + runOnlyOnProcessStart +
                '}';
    }
}

File: src/main/kotlin/me/oriharel/machinery/upgrades/SpeedUpgrade.kt
package me.oriharel.machinery.upgrades

import me.oriharel.machinery.Machinery
import me.oriharel.machinery.machine.MachineResourceGetProcess
import org.bukkit.Bukkit
import org.bukkit.configuration.file.YamlConfiguration
import java.util.AbstractMap.SimpleEntry
import java.util.function.Function
import java.util.stream.Collectors
import javax.script.ScriptEngineManager
import javax.script.ScriptException

class SpeedUpgrade(level: Int) : AbstractUpgrade(level) {
    protected override var costs: Map<Int, Int>?
    override fun applyUpgradeModifier(mineProcess: MachineResourceGetProcess) {
        val config: YamlConfiguration = Machinery.Companion.getInstance().getFileManager().getConfig("upgrades.yml").get()

        // period is in ticks. 20 ticks is a second
        val basePeriod = config.getInt("speed.basePeriod") * 20
        if (level == 1) { // if level 1 don't apply any calculations on base speed
            mineProcess.minePeriod = basePeriod
            return
        }
        val mgr = ScriptEngineManager()
        val engine = mgr.getEngineByName("JavaScript")
        try {
            mineProcess.minePeriod = engine.eval(config.getString("speed.calculate")!!.replace("\\{basePeriod}".toRegex(), basePeriod.toString()).replace(
                    "\\{level}".toRegex(), level.toString())) as Int
        } catch (e: ScriptException) {
            Bukkit.getLogger().severe("Failed to calculate mine period!")
            Bukkit.getLogger().severe(e.message)
        }
    }

    override val upgradeName: String
        get() = "Speed Upgrade"

    override fun getCosts(): Map<Int, Int>? {
        if (costs == null) costs = configLoad!!.getConfigurationSection("speed.costs")!!.getValues(false).entries.stream().map { entry: Map.Entry<String?, Any> -> SimpleEntry<Int, Int>(Integer.valueOf(entry.key), entry.value as String?. toInt ()) }.collect(Collectors.toMap(Function<SimpleEntry<Int, Int>, Int> { java.util.Map.Entry.key }, Function<SimpleEntry<Int, Int>, Int> { java.util.Map.Entry.value }))
        return costs
    }

    override val upgradeType: UpgradeType
        get() = UpgradeType.SPEED

    override fun toString(): String {
        return "SpeedUpgrade{" +
                "level=" + level +
                ", runOnlyOnProcessStart=" + runOnlyOnProcessStart +
                '}'
    }

    init {
        runOnlyOnProcessStart = true
        costs = null
    }
}